### PR TITLE
#181414758 feature(login):create user login functionality

### PR DESCRIPTION
**Pull request by Procech20**
### What does this PR do?
Create login functionality on the app.

### Description of Task to be completed?
- Create user login route
- Tested the route
- Documented the route

### How should this be manually tested?
To test the changes manually:
* clone repo on your local machine
* switch to develop branch  by typing the command ```git checkout ft-login-181414758```
* Run: ```yarn install```: to install all dependencies
* Create .env file
* Create a ```.env``` and the variables as mentioned in the ```.env.example``` file.
* Open the host address with the endpoint ``/api/v2/auth/login`` using an HTTP client such as ``postman``` or any other.
* Make a ``POST`` request using an ``email`` and ``password``.

### What to expect after running the commands:
* _On successful login, A user will get a message that says ``User login successful`` with a status of ``200``, and will get a token to be used to make other requests to other endpoints that requires sign in_

* _If the user doesn't exist, you'll a message that says: ``Invalid Email or password`` with the status of: ``404``._

### Any background context you want to provide?
- N/A

``` Useful links ```:
- N/A

### What are the relevant pivotal tracker stories?
* https://www.pivotaltracker.com/story/show/181414758

### Screenshots (if appropriate)
- N/A

### Questions:
- N/A

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -20,13 +20,16 @@
...
+    "import/prefer-default-export": 0,
     "import/no-dynamic-require": 0,
     "import/no-unresolved": [2, { "commonjs": true }],
     "new-cap": 0,
     "no-console": 0,
+    "no-named-as-default": 0,
     "no-param-reassign": 0,
     "no-shadow": [2, { "allow": ["req", "res", "err"] }],
     "no-undef": 0,
+    "no-underscore-dangle": 0,
```
Comment (didasmbalanya) [inline .eslintrc.json]: Updating rules to suit your style is not efficient. These rules are meant to reserve code quality. Please undo this and follow the lining rules

File: src/utils/errorResponse.js [hunk 2]
```diff
@@ -0,0 +1,6 @@
+export default (res, status, message) => {
+  res.status(status).json({
+    success: false,
+    message,
```
Comment (didasmbalanya) [inline src/utils/errorResponse.js]: the message key should be a string of the error message, and also add error which should be an object in case the whole error object needs to be sent to the client side

File: src/controllers/user.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-import bcrypt from 'bcrypt';
+import bcrypt from 'bcryptjs';
```
Comment (didasmbalanya) [inline src/controllers/user.js]: did you have any conversation with your team on why you switched to bcrypt js? Please do this on slack and agree if its the best way forward. Support your decision to do so

File: src/database/config/config.js [hunk 4]
```diff
@@ -3,30 +3,12 @@ require('dotenv').config();
...
 };
 export const production = {
   url: process.env.DB_CONNECT,
   dialect: 'postgres',
-  dialectOptions: {
-    ssl: {
-      require: true,
-      rejectUnauthorized: false,
-    },
-  },
```
Comment (didasmbalanya) [inline src/database/config/config.js:31]: Good job on noticing the replication

File: src/helpers/validator.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import Joi from 'joi';
+
+const Schema = {
+  login: Joi.object().keys({
+    email: Joi.string()
+      .email()
+      .required()
+      .error(new Error('Please fill all fields')),
```
Comment (didasmbalanya) [inline src/helpers/validator.js]: This error message won't work for a non email field

File: src/helpers/validator.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    password: Joi.string()
+      .required()
+      .trim()
+      .error(new Error('Please fill all fields')),
```
Comment (didasmbalanya) [inline src/helpers/validator.js:9]: add more validation, Capital, number, and length. As well as handle the error message correctly

File: src/swagger/response.js [hunk 7]
```diff
@@ -1,28 +0,0 @@
...
-  },
-  409: {
-    description: 'Conflicts',
-  },
-  500: {
-    description: 'Server Error',
-  },
-};
-
-export default responses;
```
Comment (didasmbalanya) [inline src/swagger/response.js:28]: Why was this removed?

**Review by didasmbalanya — CHANGES_REQUESTED** [review]
Good work so far, a few more things to work on though

File: src/helpers/validator.js (hunk 6 above)
Comment (Procech20) [reply src/helpers/validator.js:9]: These were handled on user signup, I thought I only had to validate if the password a user puts in matches the one he used on signup

File: src/swagger/response.js (hunk 7 above)
Comment (Procech20) [reply src/swagger/response.js:28]: I was told to use #27 documentation format and it did not use this. So I removed it.

File: src/database/config/config.js (hunk 4 above)
Comment (Procech20) [reply src/database/config/config.js:31]: Thanks

File: src/controllers/user.js (hunk 3 above)
Comment (Procech20) [reply src/controllers/user.js]: okay

**Comment by Robertishimwe** [comment]
you have to set token in the header after successful login

File: package.json [hunk 8]
```diff
@@ -1,75 +1,75 @@
...
+  "author": "Code Ninjas from Andela ATLP-6",
+  "license": "MIT",
+  "scripts": {
+    "init-dir": "sequelize init",
+    "migrate": "yarn sequelize-cli db:migrate",
+    "down": "yarn sequelize-cli db:migrate:undo",
+    "seed": "yarn sequelize-cli db:seed:all",
+    "unseed": "yarn sequelize-cli db:seed:undo:all",
+    "build": "babel src -d lib",
+    "start": "cross-env NODE_ENV=production yarn run build && node lib/index.js",
```
Comment (niyojuste) [inline package.json]: Verify this script, please. You are setting NODE_ENV for building instead of running the index file, hence, the environment is never set to 'production'.
Should be `"start": "yarn run build && cross-env NODE_ENV=production node lib/index.js"`

**Review by niyojuste — CHANGES_REQUESTED** [review]

File: src/controllers/auth.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+const { findUser } = userServices;
+const { signToken, checkPassword } = encryption;
+
+class Auth {
+  static async login(req, res, next) {
+    try {
+      const { email, password } = req.body;
+      const foundUser = await findUser({ email });
+      if (!foundUser) return ErrorResponse(res, 404, 'User Not found');
```
Comment (ericus123) [inline src/controllers/auth.js]: Don't return **user not found**.
For security purposes please just say **Incorrect credentials**

**Review by ericus123 — CHANGES_REQUESTED** [review]

**Review by ericus123 — APPROVED** [review]

**Review by niyojuste — APPROVED** [review]
